packages/workspace/src/node/default-workspace-server.ts

```diff
@@ -80,12 +80,37 @@ export class DefaultWorkspaceServer implements WorkspaceServer {
 
     async setRoot(uri: string): Promise<void> {
         this.root = new Deferred();
+        const listUri: string[] = [];
+        const oldListUri = await this.getRecentRoots();
+        listUri.push(uri);
+        if (oldListUri) {
+            oldListUri.forEach(element => {
+                // Store the new recentRoots as the first one.
+                // Remove duplicate
+                if (element !== uri && element.length > 0) {
+                    listUri.push(element);
+                }
+            });
+        }
         this.root.resolve(uri);
         this.writeToUserHome({
-            recentRoots: [uri]
+            recentRoots: listUri
         });
     }
 
+    async getRecentRoots(): Promise<string[]> {
+        const listUri: string[] = [];
+        const data = await this.readFromUserHome();
+        if (data && data.recentRoots) {
+            data.recentRoots.forEach(element => {
+                if (element.length > 0) {
```

Comment:
If we check the `element.length` (to avoid inconsistent entries) why do not we check the existence of the old workspace roots and filter out those that get deleted?

Reply:
Here I check for the empty string. It occurs after the workspace is closed and  an empty entry in ".theia/recentworkspace.json" is created. We don't need to put an entry in the selection list for a closed workspace.

Reply:
> empty entry in ".theia/recentworkspace.json" is created

I need to verify this during the review. Thanks!

Reply:
> why do not we check the existence of the old workspace roots and filter out those that get deleted

We could also or instead filter the list as it's prepared to be saved to disk in ```setWorkspace()```. I think we need to do both to cover all cases (e.g. if a workspace on the list was deleted since opening the last workspace, recentworkspace.json would still have it as an element).